sql_metadata/parser.py

```diff
@@ -0,0 +1,508 @@
+            ):
+                subqueries_names.append(str(token))
+
+        self._subqueries_names = subqueries_names
+        return self._subqueries_names
+
+    @property
+    def values(self) -> List:
+        """
+        Returns list of values from insert queries
+        """
+        if self._values:
+            return self._values
+        values = []
+        for token in self.tokens:
+            if (
+                token.last_keyword_normalized == "VALUES"
+                and token.is_in_parenthesis
+                and token.next_token.is_punctuation
+            ):
+                if token.is_integer:
+                    value = int(token.value)
+                elif token.is_float:
+                    value = float(token.value)
+                else:
+                    value = token.value.strip("'\"")
+                values.append(value)
+        self._values = values
+        return self._values
+
+    @property
+    def values_dict(self) -> Dict:
+        """
+        Returns dictionary of column-value pairs.
+        If columns are not set the auto generated column_<col_number> are added.
+        """
+        values = self.values
+        if self._values_dict or not values:
+            return self._values_dict
+        columns = self.columns
+        if not columns:
+            columns = [f"column_{ind + 1}" for ind in range(len(values))]
+        values_dict = dict(zip(columns, values))
+        self._values_dict = values_dict
+        return self._values_dict
+
+    @property
+    def comments(self) -> List[str]:
+        """
+        Return comments from SQL query
+
+        :rtype: List[str]
+        """
+        return Generalizator(self._raw_query).comments
+
+    @property
+    def without_comments(self) -> str:
+        """
+        Removes comments from SQL query
+
+        :rtype: str
+        """
+        return Generalizator(self._raw_query).without_comments
+
+    @property
+    def generalize(self) -> Optional[str]:
+        """
+        Removes most variables from an SQL query
+        and replaces them with X or N for numbers.
+
+        Based on Mediawiki's DatabaseBase::generalizeSQL
+
+        :rtype: Optional[str]
+        """
+        return Generalizator(self._raw_query).generalize
+
+    def _add_to_columns_subsection(self, keyword: str, column: str):
+        sections = {
+            "SELECT": "select",
+            "WHERE": "where",
+            "ORDERBY": "order_by",
+            "ON": "join",
+            "INTO": "insert",
+            "SET": "update",
+        }
+        section = sections[keyword]
+        self._columns_dict = self._columns_dict or dict()
+        self._columns_dict.setdefault(section, UniqueList()).append(column)
+
+    def _preprocess_query(self) -> Optional[str]:
+        """
+        Perform initial query cleanup
+
+        :rtype str
+        """
+        # 0. remove newlines
+        if self._raw_query is None:
+            return None
```

Comment:
`self._raw_query` can no longer by `None`.

```suggestion
    def _preprocess_query(self) -> str:
        """
        Perform initial query cleanup

        :rtype str
        """
        if self._raw_query == "":
            return ""

        # 0. remove newlines
```

So just leave the empty string check to leave early.

Reply:
And we may simply pass the SQL query to preprocess as an argument here, making this method a static one. This will make it easy to unit test this method.

But I'll deal with that later on - let's merge it as it is now 🙂